### PR TITLE
feat: Add RBF and CPFP Batcher wallets

### PR DESCRIPTION
**Pull request by yash1io**
- Adds batcher wallet
- Implements RBF and CPFP strategies

## Summary by CodeRabbit

- **New Features**
	- Introduced a sophisticated batching mechanism for Bitcoin transactions, allowing for improved efficiency and flexibility.
	- Enhanced wallet capabilities with new interfaces and methods for lifecycle management, fee strategies, and transaction batching.
	- Implemented a caching mechanism for managing transaction batches, improving transaction processing and storage efficiency.

- **Bug Fixes**
	- Resolved formatting issues in the workflow configuration to enhance clarity and adherence to standards.
	- Improved error handling and validation in wallet transaction processing.

- **Tests**
	- Added extensive test suites for RBF and CPFP functionalities, ensuring reliability in transaction processing and fee adjustments.
	- Enhanced wallet tests with better structure, error handling, and transaction assertions.
	- Introduced a comprehensive suite of tests for the caching mechanism, validating batch operations and request management.

### CODE REVIEW

**Comment by yash1io** [comment]
@coderabbitai review

File: btc/wallet.go [hunk 1]
```diff
@@ -801,3 +807,44 @@ func calculateTotalSendAmount(req []SendRequest) int64 {
...
+	}
+	sacpTx := btcTx.MsgTx()
+	err = validateSacp(sacpTx)
+	if err != nil {
+		return nil, err
+	}
+	return sacpTx, nil
+}
+
+func validateSacp(tx *wire.MsgTx) error {
```
Comment (Revantark) [inline btc/wallet.go:839]: Can you move these errors to their own variables like ErrSACPNoInputs..etc

File: btc/batcher.go [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+		return err
+	}
+
+	totalSendAmt := calculateTotalSendAmount(sends)
+
+	if walletBalance+spendsAmount < totalSendAmt {
+		return ErrBatchParametersNotMet
+	}
+
+	return validateRequests(spends, sends, sacps)
```
Comment (Revantark) [inline btc/batcher.go:448]: I think this check should happen first as it doesn't call external APIs.

File: btc/batcher.go [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+)
+
+var (
+	SegwitSpendWeight int = txsizes.RedeemP2WPKHInputWitnessWeight
+)
+var (
+	ErrBatchNotFound             = errors.New("batch not found")
+	ErrBatcherStillRunning       = errors.New("batcher is still running")
+	ErrBatcherNotRunning         = errors.New("batcher is not running")
+	ErrBatchParametersNotMet     = errors.New("batch parameters not met")
```
Comment (Revantark) [inline btc/batcher.go:28]: The error is generic and does not specify what exactly went wrong. Maybe you can break the error into multiple errors and use them.

File: btc/batcher.go [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+	return w.sw.SignSACPTx(tx, idx, amount, leaf, scriptAddr, witness)
+}
+
+// Send creates a batch request , saves it in the cache and returns a tracking id
+func (w *batcherWallet) Send(ctx context.Context, sends []SendRequest, spends []SpendRequest, sacps [][]byte) (string, error) {
+	if err := w.validateBatchRequest(ctx, spends, sends, sacps); err != nil {
+		return "", err
+	}
+
+	id := chainhash.HashH([]byte(fmt.Sprintf("%v_%v", spends, sends))).String()
```
Comment (Revantark) [inline btc/batcher.go:257]: If spends and sends are same for two separate requests, id will be same right ?

File: btc/batcher.go [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+		w.runPTIBatcher(ctx)
+	default:
+		panic("strategy not implemented")
+	}
+}
+
+// PTI stands for Periodic time interval
+//  1. It creates a batch at regular intervals
+//  2. It also updates the fee rate at regular intervals
+//     if fee rate increases more than threshold and there are
```
Comment (Revantark) [inline btc/batcher.go:341]: These comments are not formatted well.

File: btc/batcher.go [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+//     no batches to create
+func (w *batcherWallet) runPTIBatcher(ctx context.Context) {
+	w.wg.Add(1)
+	go func() {
+		defer w.wg.Done()
+		for {
+			ticker := time.NewTicker(w.opts.PTI)
```
Comment (Revantark) [inline btc/batcher.go:348]: `ticker` should be created before the loop.

File: btc/cpfp.go [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+			sendRequests = append(sendRequests, req.Sends...)
+			sacps = append(sacps, req.SACPs...)
+			reqIds[req.ID] = true
+		}
+
+		return spendRequests, sendRequests, sacps, reqIds
+	}()
+
+	// Return error if no requests found
+	if len(sendRequests) == 0 && len(spendRequests) == 0 {
```
Comment (Revantark) [inline btc/cpfp.go:57]: handle SACP requests

File: btc/cpfp.go [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+}
+
+// calculateFeeStats calculates the fee stats based on the required fee rate
+func calculateFeeStats(reqFeeRate int, batches []Batch) FeeStats {
+	maxFeeRate := int(0)
+	totalSize := int(0)
+	feeDelta := int(0)
+
+	for _, batch := range batches {
+		size := batch.Tx.Weight / 4
```
Comment (Revantark) [inline btc/cpfp.go:461]: consider using `blockchain.WitnessScaleFactor` to maintain consistency.

File: btc/cpfp.go [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+	txb := btcutil.NewTx(tx)
+	trueSize := mempool.GetTxVirtualSize(txb)
+
+	// Estimate the new fee
+	newFeeEstimate := (int(trueSize) * (feeRate)) + feeOverhead + bufferFee
+
+	// If the new fee estimate exceeds the current fee, rebuild the CPFP transaction
+	if newFeeEstimate > fee+feeOverhead {
+		buf := bytes.NewBuffer(make([]byte, 0, tx.SerializeSize()))
+		err = tx.Serialize(buf)
```
Comment (Revantark) [inline btc/cpfp.go:358]: Can you extract getting bytes from `wire.MsgTx` into a separate function and use it everwhere.

File: btc/cpfp.go [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+	if len(sendRequests) == 0 && len(spendRequests) == 0 {
+		return ErrBatchParametersNotMet
+	}
+
+	// Fetch fee rates and select the appropriate fee rate based on the wallet's options
+	feeRates, err := w.feeEstimator.FeeSuggestion()
+	if err != nil {
+		return err
+	}
+	requiredFeeRate := selectFee(feeRates, w.opts.TxOptions.FeeLevel)
```
Comment (Revantark) [inline btc/cpfp.go:66]: this piece of code can go just above `getFeeSats`

File: btc/cpfp.go [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+	err = withContextTimeout(c, 5*time.Second, func(ctx context.Context) error {
+		return w.cache.UpdateBatchStatuses(ctx, confirmedTxs, true, CPFP)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Calculate fee stats based on the required fee rate
+	feeStats, err := getFeeStats(requiredFeeRate, pendingBatches, w.opts)
+	if err != nil && !errors.Is(err, ErrFeeUpdateNotNeeded) {
```
Comment (Revantark) [inline btc/cpfp.go:93]: This error (ErrFeeUpdateNotNeeded) is not being returned in `getFeeStats`

File: btc/cpfp.go [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/mempool"
+	"github.com/btcsuite/btcd/wire"
+	"go.uber.org/zap"
+)
+
+type FeeStats struct {
+	MaxFeeRate int
+	TotalSize  int
```
Comment (Revantark) [inline btc/cpfp.go:19]: `TotalSize` is not being used anywhere.

File: btc/cpfp.go [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+	err = withContextTimeout(c, 5*time.Second, func(ctx context.Context) error {
+		utxos, err = w.indexer.GetUTXOs(ctx, w.address)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	// Verify CPFP conditions
+	if err := verifyCPFPConditions(utxos, pendingBatches, w.address); err != nil {
```
Comment (Revantark) [inline btc/cpfp.go:210]: can you comment this if `VerifyCPFPConditions` is not ready yet ?

File: btc/batcher.go [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+							w.logger.Error("failed to update fee rate", zap.Error(err))
+						} else {
+							w.logger.Info("fee update skipped")
+						}
+					} else {
+						w.logger.Info("batch fee updated", zap.String("strategy", string(w.opts.Strategy)))
+					}
+				} else {
+					w.logger.Info("new batch created", zap.String("strategy", string(w.opts.Strategy)))
+				}
```
Comment (Revantark) [inline btc/batcher.go:373]: `createBatch` might fail for different reasons. Calling `updateFeeRate` on error might lead to corruption of data.

File: btc/batcher.go [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+
+// updateFeeRate updates the fee rate based on the strategy
+func (w *batcherWallet) updateFeeRate() error {
+	feeRates, err := w.feeEstimator.FeeSuggestion()
+	if err != nil {
+		return err
+	}
+	requiredFeeRate := selectFee(feeRates, w.opts.TxOptions.FeeLevel)
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (Revantark) [inline btc/batcher.go:389]: Everywhere parent context is being considered. You can pass the parent context here too if possible.

**Comment by Revantark** [comment]
In all the timeout contexts, `5s` is being used. You can make it a global constant and use it.

File: btc/rbf.go [hunk 16]
```diff
@@ -0,0 +1,680 @@
...
+			zap.Any("avoidUtxos", avoidUtxos),
+			zap.Uint("fee", fee),
+			zap.Int("requiredFeeRate", feeRate),
+			zap.Bool("checkValidity", checkValidity),
+			zap.Int("depth", depth),
+		)
+		return nil, nil, nil, ErrBuildRBFDepthExceeded
+	}
+
+	var sacpsIn int
```
Comment (Revantark) [inline btc/rbf.go:380]: totalSACPInAmount should be the right name

**Review by Revantark — APPROVED** [review]

**Pull request merged**